### PR TITLE
Begin warning people about spaces in model names

### LINKED ISSUE #9397: [CT-3564] for deprecation - we should not support spaces in model names
### Housekeeping

- [X] I am a maintainer of dbt-core

### Short description

Spaces in model names was only ever supported incidentally & by accident. It’s never been documented. 
- Having a space in a model name makes in impossible to select via `--select` on the model name because the argument gets split into pieces over spaces very early in the pipeline.
- Most warehouses will not like a table, or other object, with a space in its name.
- We may have to support table/view names with spaces for completeness on some platforms, but I we shouldn't support them in dbt's internal identifiers.

### Acceptance criteria

We should mark spaces in model names:
- for deprecation (raise warning) in v1.8 
- fully remove support in v1.9 (raise error)

### Suggested Tests

1. test model name with space returns a warning in v1.8, error in v1.9

### Impact to Other Teams

Impact: Exp and Adapters

### Will backports be required?

No

### Context

Impacting IDE team and file system. 
How many customers are experiencing this warning?

### PR DESCRIPTION
**Pull request by QMalcolm**
resolves #9397

### Problem

We don't support models with spaces in their names. However, we haven't been actually enforcing this. If a person have spaces in their model names, it causes issues when using selectors. Additionally, depending on a person's operating system, there were other edge case problems with spaces in model names.

### Solution

Begin warning people about spaces in their model names. Depending on some flags, we give more or less information. If `debug` mode _is not on_, then we log out the first offending model name and a count of how many offending model names their are in total. If `debug` mode _is on_, then we log out every offending model name. Additionally, by default currently these logs are warnings and won't stop dbt from running. However, if one sets `allow_spaces_in_model_names` to `False` in their `dbt_project.yml`, then the logs become errors and found offending model names will stop dbt from running.

### Checklist

- [X] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [X] I have run this code in development and it appears to resolve the stated issue  
- [X] This PR includes tests, or tests are not required/relevant for this PR
- [X] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX
- [X] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions

### CODE REVIEW

**Review by emmyoop — CHANGES_REQUESTED** [review]

File: core/dbt/events/types.py
Comment (QMalcolm) [reply core/dbt/events/types.py:416]: We can do that! I originally made it `WARNLevel` as that should be the initial default. But with the way that this is implemented, making it dynamic makes sense 👍

File: core/dbt/parser/manifest.py
Comment (QMalcolm) [reply core/dbt/parser/manifest.py:664]: I don't think we want to immediately quit, it results in a worse experience and isn't actually necessary. By not immediately quitting we allow people to get either the total count of model names with issues OR all the model names with issues if using `--debug`. If there are 47 bad model names in someone's project and we quit after detecting the first one, then we push the person towards a much slower and frustrating process of fixing one and then running again only to discover another problem name (again and again and again).

**Comment by QMalcolm** [comment]
I just checked out main and am still running into the `AttributeError: 'Namespace' object has no attribute 'SEND_ANONYMOUS_USAGE_STATS'` error in those tests. So something with my local environment is screwy and not getting to the error we're seeing in the GHA. Gonna nuke my tox cache locally and see if that fixes things.

**Comment by QMalcolm** [comment]
After busting my local tox cache, running the tests via `make test` they all pass. However if I use pytest to directly test the failing graph tests (`pytest tests/unit/test_graph.py`) they fail with the `AttributeError: 'Namespace' object has no attribute 'SEND_ANONYMOUS_USAGE_STATS'`.

**Comment by QMalcolm** [comment]
We had a fix on main for the failing integration tests in `test_pp_vars.py`. Rebasing to bring in those changes.

File: core/dbt/events/types.py
Comment (emmyoop) [inline core/dbt/events/types.py:16]: Are you going to do this TODO before merging?

File: tests/functional/manifest_validations/test_check_for_spaces_in_model_names.py
Comment (emmyoop) [inline tests/functional/manifest_validations/test_check_for_spaces_in_model_names.py:12]: ❤️

File: core/dbt/events/types.py
Comment (emmyoop) [inline core/dbt/events/types.py:445]: Since this event fires when there are a non-zero number of models with spaces you need to account for there being only 1 model in how the sentence is structured.  ie "Found 1 model..." instead of "Found 1 models..."

**Review by emmyoop — CHANGES_REQUESTED** [review]

File: core/dbt/events/types.py
Comment (QMalcolm) [reply core/dbt/events/types.py:16]: I do not. That would entail opening a a PR in dbt-common and also getting it released, and I didn't want that to block this PR

Comment (QMalcolm) [reply core/dbt/events/types.py:16]: Follow up tasks assigned to me:
* https://github.com/dbt-labs/dbt-core/issues/9914
* https://github.com/dbt-labs/dbt-common/issues/107

File: core/dbt/events/types.py
Comment (QMalcolm) [reply core/dbt/events/types.py]: Restructured the whole sentence so that we could just use `model(s)` previously there were a bunch of words that would have to be interpolated depending on whether there was 1 problematic name or 2+ problematic names.

**Review by emmyoop — APPROVED** [review]
Looks good!

**Pull request merged**
